scvi/external/scar/_model.py

```diff
@@ -0,0 +1,435 @@
+        raw_adata = raw_adata[:, raw_adata.var_names.isin(adata.var_names)]
+        raw_adata = raw_adata[raw_adata.X.sum(axis=1) >= min_raw_counts]
+
+        raw_adata.obs["total_counts"] = raw_adata.X.sum(axis=1)
+
+        sample = int(sample)
+        idx = np.random.choice(
+            raw_adata.shape[0], size=min(raw_adata.shape[0], sample), replace=False
+        )
+        raw_adata = raw_adata[idx]
+        if verbose:
+            print(
+                "Randomly sample ",
+                sample,
+                " droplets to calculate the ambient profile.",
+            )
+        # initial estimation of ambient profile, will be update
+        ambient_prof = raw_adata.X.sum(axis=0) / raw_adata.X.sum()
+
+        if verbose:
+            print("Estimating ambient profile for ", feature_type, "...")
+
+        i = 0
+        while i < iterations:
+            # calculate joint probability (log) of being cell-free droplets for each droplet
+            log_prob = []
+            batch_idx = np.floor(
+                np.array(range(raw_adata.shape[0])) / raw_adata.shape[0] * n_batch
+            )
+            for b in range(n_batch):
+                try:
+                    count_batch = raw_adata[batch_idx == b].X.astype(int).A
+                except MemoryError:
+                    raise MemoryError("use more batches by setting a higher n_batch")
+                log_prob_batch = Multinomial(
+                    probs=torch.tensor(ambient_prof), validate_args=False
+                ).log_prob(torch.Tensor(count_batch))
+                log_prob.append(log_prob_batch)
+            log_prob = np.concatenate(log_prob, axis=0)
+            raw_adata.obs["log_prob"] = log_prob
+            raw_adata.obs["droplets"] = "other droplets"
+            # cell-containing droplets
+            raw_adata.obs.loc[
+                raw_adata.obs_names.isin(adata.obs_names), "droplets"
+            ] = "cells"
+            # identify cell-free droplets
+            raw_adata.obs["droplets"] = raw_adata.obs["droplets"].mask(
+                raw_adata.obs["log_prob"] >= np.log(prob) * raw_adata.shape[1],
+                "cell-free droplets",
+            )
+            emptydrops = raw_adata[raw_adata.obs["droplets"] == "cell-free droplets"]
+            if emptydrops.shape[0] < 50:
+                raise Exception("Too few emptydroplets! Lower the prob parameter")
+            ambient_prof = emptydrops.X.sum(axis=0) / emptydrops.X.sum()
+            i += 1
+            if verbose:
+                print("iteration: ", i)
+
+        # update ambient profile
+        for ft in feature_type:
+            tmp = emptydrops[:, emptydrops.var["feature_types"] == ft]
+            adata.uns[f"ambient_profile_{ft}"] = pd.DataFrame(
+                tmp.X.sum(axis=0).reshape(-1, 1) / tmp.X.sum(),
+                index=tmp.var_names,
+                columns=[f"ambient_profile_{ft}"],
+            )
+
+        if kneeplot:
```

Comment:
we make it a point to not include any plotting code in this package, so I would prefer to remove this (as well as in the docstring)

Reply:
sounds good, I removed it in the latest commit